File: services/customers-service/src/test/java/br/com/deltaglobalbank/customers/domain/customer/valueobjects/BirthDateTests.java
package br.com.deltaglobalbank.customers.domain.customer.valueobjects;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.time.LocalDate;

import org.junit.jupiter.api.Test;

class BirthDateTests {

    @Test
    void mustAcceptAnAdult() {
        assertDoesNotThrow(() -> new BirthDate(LocalDate.of(1990, 5, 20)));
    }

    @Test
    void mustAcceptSomeoneWhoJustTurned18() {
        assertDoesNotThrow(() -> new BirthDate(LocalDate.now().minusYears(18)));
    }

    @Test
    void mustRejectSomeoneUnder18() {
        assertThrows(IllegalArgumentException.class,
            () -> new BirthDate(LocalDate.now().minusYears(18).plusDays(1)));
    }

    @Test
    void mustRejectAFutureDate() {
        assertThrows(IllegalArgumentException.class, () -> new BirthDate(LocalDate.now().plusDays(1)));
    }
}

File: services/customers-service/src/test/kotlin/br/com/deltaglobalbank/customers/domain/customer/valueobjects/BirthDateTests.kt
package br.com.deltaglobalbank.customers.domain.customer.valueobjects

import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertThrows
import java.time.LocalDate
import org.junit.jupiter.api.Test

class BirthDateTests {

    @Test
    fun `must accept an adult`() {
        assertDoesNotThrow { BirthDate(LocalDate.of(1990, 5, 20)) }
    }

    @Test
    fun `must accept someone who just turned 18`() {
        assertDoesNotThrow { BirthDate(LocalDate.now().minusYears(18)) }
    }

    @Test
    fun `must reject someone under 18`() {
        assertThrows(IllegalArgumentException::class.java) {
            BirthDate(LocalDate.now().minusYears(18).plusDays(1))
        }
    }

    @Test
    fun `must reject a future date`() {
        assertThrows(IllegalArgumentException::class.java) { BirthDate(LocalDate.now().plusDays(1)) }
    }
}